runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,481 @@
+// to have observed before the data can be deleted. By making
+// use of the fact that (barring integer overflow) the largest
+// targetEpoch is always pushed as the head, we have a relative
+// sorting order. If we ask for some targetEpoch X, and for some
+// node N we have X >= N.targetEpoch, then we can claim N and
+// N.next and so and so forth since N.targetEpoch >= N.next.targetEpoch. 
+struct defer_node {
+  uint64_t targetEpoch;
+  void *data;
+  int numData;
+  struct defer_node *next;
+};
+
+// Thread-local data that keeps track of our observed global state.
+struct tls_node {
+  /*
+    The number of checkpoints passed through
+  */
+  uint64_t passedCheckpoints;
+  /*
+    The current observed epoch. This is used to ensure that writers
+    do not delete an older instance that corresponds to this epoch.
+  */
+  atomic_uint_least64_t epoch;
+  /*
+    Number of tasks that this thread has running. Required so
+    that threads without tasks (and no longer are able to call
+    checkpoints) are not taken into account.
+  */
+  atomic_uint_least64_t nTasks;
+  /*
+    List of objects that are queued for deferred deletion, sorted by target_epoch. 
+    We defer deletion if we find that while scanning the list of all thread-specific
+    data, they are not up-to-date on that epoch. By sorting by target_epoch, we can
+    make shortcuts by removing a large bulk of data at once. The list is protected
+    by a lightweight spinlock, but is acquired using pseudo-TestAndSet operations
+    due to low contention.
+  */
+  atomic_uint_least64_t spinlock;
+  struct defer_node *deferList;
+
+  /*
+    Recycled defer nodes that can be reused.
+  */
+  struct defer_node *recycleList;
+  struct tls_node *next;
+};
+
+// List of thread-local data.
+static struct tls_node *chpl_qsbr_tls_list;
+static CHPL_TLS_DECL(struct tls_node *, chpl_qsbr_tls);
+
+static inline struct tls_node *get_tls_list(void);
+static inline struct tls_node *get_tls_list(void) {
+  return (struct tls_node *) atomic_load_uintptr_t((uintptr_t *) &chpl_qsbr_tls_list);
+}
+
+static inline uint64_t get_epoch(struct tls_node *node);
+static inline uint64_t get_epoch(struct tls_node *node) {
+  return atomic_load_uint_least64_t(&node->epoch);
+}
+
+static inline void set_epoch(struct tls_node *node, uint64_t epoch);
+static inline void set_epoch(struct tls_node *node, uint64_t epoch) {
+  return atomic_store_uint_least64_t(&node->epoch, epoch);
+}
+
+static inline uint64_t get_global_epoch(void);
+static inline uint64_t get_global_epoch(void) {
+  return atomic_load_uint_least64_t(&global_epoch);
+}
+
+static inline uint64_t advance_global_epoch(void);
+static inline uint64_t advance_global_epoch(void) {
+  return atomic_fetch_add_uint_least64_t(&global_epoch, 1);
+}
+
+static inline void observe_epoch(struct tls_node *node);
+static inline void observe_epoch(struct tls_node *node) {
+  atomic_store_uint_least64_t(&node->epoch, get_global_epoch());
```

Comment:
I think it's important that this operation in particular reflect a "full memory fence" and would appreciate a comment to that effect. (It already does have one - I just don't want somebody "optimizing" later to forget that it needs one). Otherwise, you might have stores that are in-flight even after we've observed the epoch in the checkpoint function.

Reply:
Could you add the comment like "// Full memory fence required here and is implied by the below atomic_store"?